include/git2/sys/repository.h

```diff
@@ -25,11 +25,25 @@ GIT_BEGIN_DECL
  * Note that this is only useful if you wish to associate the repository
  * with a non-filesystem-backed object database and config store.
  *
+ * Caveats: since this repository has no physical location, some systems
+ * can fail to function properly: locations under $GIT_DIR, $GIT_COMMON_DIR,
+ * or $GIT_INFO_DIR are impacted.
+ *
  * @param out The blank repository
  * @return 0 on success, or an error code
  */
 GIT_EXTERN(int) git_repository_new(git_repository **out);
 
+/**
+ * Set or unset a repository's path
+ *
+ * @param repo A new repo to set the path of.
+ * @param path The path to use.
+ *
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_repository_set_path(git_repository *repo, const char *path);
```

Comment:
That allows giving a repository a filesystem location. It's not quite clear what will happen to it, but at least that should fix the "cannot exist in repository", if a user choses to do so.

Reply:
👍 on the naming issue, and I can take care of the deprecation here if we decide for it.

Reply:
Mh. Dunno -- I think it would be easier to do in another PR to keep the code churn low and keep this reviewable. The deprecation things always require to make a sweep through all of our code, and I bet that `git_repository_path` is used in quite a lot of places.